src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,342 @@
+
+### Back to `GetValue`
+
+Let's look at how `GetValue` is defined:
+
+> [`GetValue ( V )`](https://tc39.es/ecma262/#sec-getvalue)
+>
+> 1. `ReturnIfAbrupt(V)`.
+> 2. If `Type(V)` is not `Reference`, return `V`.
+> 3. Let `base` be `GetBase(V)`.
+> 4. If `IsUnresolvableReference(V)` is `true`, throw a `ReferenceError` exception.
+> 5. If `IsPropertyReference(V)` is `true`, then
+> a. If `HasPrimitiveBase(V)` is `true`, then
+> i. Assert: In this case, `base` will never be `undefined` or `null`.
+> ii. Set `base` to `! ToObject(base)`.
+> b. Return `? base.[[Get]](GetReferencedName(V), GetThisValue(V))`.
+> 6. Else,
+> a. Assert: `base` is an Environment Record.
+> b. Return `? base.GetBindingValue(GetReferencedName(V), IsStrictReference(V))`
+
+The reference in our example is `o2.foo` which is a property reference. So we take branch 5. We don't take the branch in 5 a, since the base (`o2`) is not a primitive value (a Boolean, String, Symbol, BigInt or Number).
+
+Then we call `[[Get]]` in step 5 b. The `Receiver` we pass is `GetThisValue(V)`. In this case, it's just the basevalue of the Reference:
+
+> [`GetThisValue( V )`](https://tc39.es/ecma262/#sec-getthisvalue)
+>
+> 1. Assert: `IsPropertyReference(V)` is `true`.
+> 2. If `IsSuperReference(V)` is `true`, then
+> a. Return the value of the `thisValue` component of the reference `V`.
+> 3. Return `GetBase(V)`.
+
+For `o2.foo`, we don't take the branch in step 2, since it's not a super reference (such as `super.foo`), but we take step 3 and return the base value of the reference which is `o2`.
+
+Piecing everything together, we find out that we set the `Receiver` to be the base of the original reference, and then we keep it unchanged during the prototype chain walk. Finally, if the property we find is an accessor property, we use the `Receiver` as the **this value** when calling it.
+
+In particular, the **this value** inside a getter refers to the original object where we tried to get the property from, not the one where we found the property during the prototype chain walk.
+
+Let's try it out!
+
+```javascript
+const o1 = { x: 10, get foo() { return this.x; } };
+let o2 = {};
+Object.setPrototypeOf(o2, o1);
+o2.x = 50;
+o2.foo; // will return 50
+```
+
+In this example, we have an accessor property called `foo` and we define a getter for it. The getter returns `this.x`.
+
+Then we access `o2.foo` - what does the getter return?
+
+We found out that when we call the getter, the **this value** is the object where we originally tried to get the property from, not the object where we found it. In this case the **this value** is `o2`, not `o1`. We can verify that by checking whether the getter returns `o2.x` or `o1.x`, and indeed, it returns `o2.x`.
+
+(Note that setting `o2.x = 50` adds a property called `x` in `o2` and doesn't overwrite the property `x` in `o1`.)
```

Comment:
Setters on the proto are actually super confusing when following through the spec. If `o1.x` was non-writable, this would error/silently nop. If `o1.x` was a setter, that would invoke the setter and not create an `o2.x`. I think for the scope of the tutorial we should remove mentions of property sets.

Reply:
Right, removed the setting, it's much simpler that way.